drivers/dai/intel/ssp/ssp.c

```diff
@@ -0,0 +1,1992 @@
+/*
+ * Copyright (c) 2022 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <spinlock.h>
+#include <devicetree.h>
+#define LOG_DOMAIN dai_intel_ssp
+#include <logging/log.h>
+LOG_MODULE_REGISTER(LOG_DOMAIN);
+
+#include "ssp.h"
+
+#define DT_DRV_COMPAT intel_ssp_dai
+
+#define dai_set_drvdata(dai, data) (dai->priv_data = data)
+#define dai_get_drvdata(dai) dai->priv_data
+#define dai_get_mn(dai) dai->plat_data.mn_inst
+#define dai_get_ftable(dai) dai->plat_data.ftable
+#define dai_get_fsources(dai) dai->plat_data.fsources
+#define dai_mn_base(dai) dai->plat_data.mn_inst->base
+#define dai_base(dai) dai->plat_data.base
+
+#define DAI_DIR_PLAYBACK	0
+#define DAI_DIR_CAPTURE		1
+#define SSP_ARRAY_INDEX(dir) dir == DAI_DIR_RX ? DAI_DIR_CAPTURE : DAI_DIR_PLAYBACK
+
+static inline void dai_mn_write(struct dai_intel_ssp *dp, uint32_t reg, uint32_t val)
+{
+	sys_write32(val, dai_mn_base(dp) + reg);
+}
+
+static inline uint32_t dai_mn_read(struct dai_intel_ssp *dp, uint32_t reg)
+{
+	return sys_read32(dai_mn_base(dp) + reg);
+}
+
+static void dai_ssp_write(struct dai_intel_ssp *dp, uint32_t reg, uint32_t val)
+{
+	sys_write32(val, dai_base(dp) + reg);
+}
+
+static uint32_t dai_ssp_read(struct dai_intel_ssp *dp, uint32_t reg)
+{
+	return sys_read32(dai_base(dp) + reg);
+}
+
+static void dai_ssp_update_bits(struct dai_intel_ssp *dp, uint32_t reg, uint32_t mask, uint32_t val)
```

Comment:
Looks like there are many functions common for all DAI-based drivers. Can we create a library or a helper instead of repeating dai_ssp_update_bit, dai_dmic_update_bit, dai_XYZ_update_bit ?

Reply:
yes there's overlap, but there are not too many. I did this when creating blob generator for alsa-utils. I would actually like to move the common part to zephyr side.

If you look at other zephyr drivers, almost all define they own bit macros, so we are not very badly aligned in that sense. So I would refactor this then, when I bring in other drivers than ssp. As said the overlap is not big and I will probably not get it right without other driver -> anyway I would need to revisit this.

Reply:
I removed now many of the extra set bits etc and used zephyr sys_read and write